### PR TITLE
[RSDK-1062] implement orb GetPointCloudMap

### PR DESCRIPTION
**Pull request by nicksanford**
[Ticket](https://viam.atlassian.net/browse/RSDK-1062)

Add GetPointCloudMap GRPC endpoint to orbslam.

Tested via:

```bash
# request new endpoint
grpcurl -d '{"name": "run-slam"}'  -plaintext -protoset <(~/slam-clone/slam-libraries/grpc/bin/buf build -o -) $(sudo netstat -tulpn | grep LISTEN | grep orb_grpc_se | awk '{print $4}' | sort | head -n 1) viam.service.slam.v1.SLAMService/GetPointCloudMap | jq -r '.pointCloudPcd' | base64 -d > get_point_cloud_map.pcd

# request old endpoint
grpcurl -d '{"name": "run-slam", "mime_type": "pointcloud/pcd"}'  -plaintext -protoset <(~/slam-clone/slam-libraries/grpc/bin/buf build -o -) $(sudo netstat -tulpn | grep LISTEN | grep orb_grpc_ | awk '{print $4}' | sort | head -n 1) viam.service.slam.v1.SLAMService/GetMap | jq -r '.pointCloud.pointCloud' | base64 -d > get_map.pcd

#convert new & old response value to ascii for easier analysis
pcl_convert_pcd_ascii_binary get_map.pcd get_map_ascii3.pcd 0
pcl_convert_pcd_ascii_binary get_point_cloud_map.pcd get_point_cloud_map_ascii3.pcd 0

# observe they are the same except for the color
vimdiff get_map_ascii3.pcd get_point_cloud_map_ascii3.pcd
```
[get_point_cloud_map_ascii3.pcd.txt](https://github.com/viamrobotics/slam/files/10520635/get_point_cloud_map_ascii3.pcd.txt)
[get_map_ascii3.pcd.txt](https://github.com/viamrobotics/slam/files/10520641/get_map_ascii3.pcd.txt)

 attached

I also confirmed that app.viam UI is able to render the new returned point cloud:

With the `dev` config feature flag off (i.e. old GetMap endpoint):
old

With the `dev` config feature flag on (i.e. old GetPointCloudMap endpoint):
new

### CODE REVIEW

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (nicksanford) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:113]: This is equal to this if statement https://github.com/viamrobotics/slam/pull/139/files#diff-b68d96da1bc970406f0f4402811d582812550e7c8450a636d73540f0363c73d8L318, just fitting it into a `GetPointCloudMapResponse` return value.

**Comment by nicksanford** [comment]
I've been told by the team that these types of GRPC methods are not tested by unit tests & therefor are not automatically tested in the slam repo itself. 

I'm curious if people would find it valuable to add integration / GRPC test to this repo? We could maybe use a python, go or any language with a GRPC client the team would find ergonomic to write tests in. Doing so would automate the manual testing I have been doing with grpcurl to verify the endpoints behave how I expect.

I'd be happy to take on that work if we would find it valuable. Also completely happy if the answer is "post modularization we can consider it".

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (tessavitabile) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:119]: [nit] no need to get currPose

**Review by tessavitabile — APPROVED** [review]
LGTM mod nit

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (nicksanford) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:133]: @JohnN193 I copied the pattern from here https://github.com/viamrobotics/slam/pull/9/files#diff-b68d96da1bc970406f0f4402811d582812550e7c8450a636d73540f0363c73d8R136 in `GetMap` & have a few questions:


1. When the original GetMap code was introduced james provided some feedback that this could be refactored to use sprintf, would this now be a good time to do that refactor?
2. Why are we creating both a `std::stringbuf` , and then creating a `std::ostream` from it & then writing to both the ostream in some parts of the code ([here](https://github.com/viamrobotics/slam/pull/139/files#diff-b68d96da1bc970406f0f4402811d582812550e7c8450a636d73540f0363c73d8R137) & to the stringbuf [here](https://github.com/viamrobotics/slam/pull/139/files#diff-b68d96da1bc970406f0f4402811d582812550e7c8450a636d73540f0363c73d8R182) in others? Ideally I'd like to only have a single variable which is written to, but please educate me if we actually need to have 2 different objects with pointers to each other for some reason. 


cc @tessavitabile for C++ expertise

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (nicksanford) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:174]: After talking with @JohnN193 it turns out that I need to remove color information from this given that orbslam doesn't have probability information & the RGB channel is currently arbitrarily representing hight.

I'm also going to try to refactor this code to have helper functions for the ascii & binary writing to the buffer & try to only use a single buffer variable for the entire function.

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (jeremyhyde-viam) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:133]: I was looking into ifstream and simialr functions while doing GetInternalState [carto] and while it may not always be the case, ostream takes in a buffer object (either a file buffer  or string buffer). To my knowledge the ostream is handling the addition of bytes to the buffer object.

You might be able to get away with just a stringbuf but the ostream helps optimize performance.

https://stackoverflow.com/questions/48728330/whats-the-mechanism-of-iostreams-with-buffer

https://cplusplus.com/reference/ostream/ostream/ostream/

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (jeremyhyde-viam) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:117]: Looks like they are used on line 150 to help calculate the color values. You can just change this to MaxInt and MinInt if you want to remove magic numbers but use the same color calc

Comment (JohnN193) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:117]: talked on this offline, these were for scaling the current height to color which is something we do not need/want anymore

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (JohnN193) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:133]: yeah refactoring here would be fine, should help clean stuff up

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (tessavitabile) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:174]: oh right, I forgot how we're using colors currently!

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (jeremyhyde-viam) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:169]: [minor/nit] can we split all of the color math to its own function? I'd love to have GetMap be as streamlined as possible for ease of reference in the future

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.h
Comment (jeremyhyde-viam) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.h:44]: Can we add comments to at the very least the GRPC functions?

**Review by jeremyhyde-viam — CHANGES_REQUESTED** [review]
Good job moving this over, glad it isnt looking as involved as cartographer's GetPointCloudMap! A few minor improvements and answers to posed questions

**Comment by tessavitabile** [comment]
By the way, just a reminder to use the imperative mood for commit messages, and make the PR title the same as the intended commit message: https://viam.atlassian.net/wiki/spaces/ENG/pages/262371/Viam+Engineering+Guidelines

**Comment by nicksanford** [comment]
> 

Thanks for the reminder!

Once this is all approved I'll squash all changes & rename the commit message to the title. I believe that will bring this PR back into compliance with your reminder but please let me know if I've missed something. 

Also, if I'm squashing on merge, do the incremental commits (which happen during debugging / responding to feedback) require adhering to the imperative mood (given that they will be squashed away on merge I believe)?

**Comment by tessavitabile** [comment]
No, incremental commits don't require imperative mood. Only the commit that gets merged.

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (nicksanford) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:150]: + a refactor to remove the color information (as it is not longer meaningful) & refactor to only use a single string type to create the pointcloud pcd, with helper functions where possible.

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (nicksanford) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:58]: I'm passing buffer as a reference as I'm under the impression (from the tour of C++ book) that passing references rather than pointers is preferable whenever possible, which it is in this case.  Please let me know if that is not the case.

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (nicksanford) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:60]: @tessavitabile could I trouble you to confirm this is the safest way to get each of the 8 bytes out of a 32 bit float & add each byte to a string? I use a `uint8_t` pointer as I'm under the impression `const char *` (the pointer type the old implementation cast to) may not always be 8 bits on all platforms, which would break the expectations of our implementation, but please correct me if there is an error in my thinking / a different way I should do this.

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (nicksanford) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc]: done

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.h
Comment (nicksanford) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.h:53]: done

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (nicksanford) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc]: team decided to remove color info all together, so I think this is now addressed.

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (tessavitabile) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:58]: I've also seen the pattern of passing a const reference (for things that can be const) and a non-const pointer (for things that cannot be const). I'm not sure why people use pointers for non-const sometimes. But I think what you've done here makes sense.

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (tessavitabile) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:60]: I'm not sure this is necessary here. We only support a couple platforms, and we test those platforms. If we were to strive for platform independence, I would want to test on more platforms (and we would also have other things to fix, such as assuming the size of a float in the pcd header). For now, I think we shouldn't make parts of the code platform-independent, if it adds complexity.

**Review by tessavitabile — CHANGES_REQUESTED** [review]
I think it's cool that you looked into this, but I don't recommend making the change for platform independence at this time.

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (tessavitabile) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:60]: We discussed offline that the motivation is readability. I don't feel that I can speak to the safest way to perform this--I would be looking it up for myself.

One suggestion I have is to add a check that sizeof(float)==4 and fail fast if this is not the case. Otherwise, LGTM

**Review by tessavitabile — APPROVED** [review]

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (jeremyhyde-viam) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:60]: For reference, Cartographer serialization uses this function:
```
void WriteSizeAsLittleEndian(uint64 size, std::ostream* out) {
  for (int i = 0; i != 8; ++i) {
    out->put(size & 0xff);
    size >>= 8;
  }
}
```
just as a reference but as Tess already pointed out I think we can focus for now on a single platform and your function can be expanded upon once we deal with more platforms.

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.h
Comment (jeremyhyde-viam) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.h:47]: [nit] remove file as PCD is a valid format to refer to a buffer and having file here might make people expect a saved file

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.h
Comment (jeremyhyde-viam) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.h:50]: [q] not sure if its worth stating that we use the right hand rule as the left hand rule is only ever used when looking at magnetic fields but clarity is never a bad thing

**Review by jeremyhyde-viam — APPROVED** [review]
LGTM, one or two nits but otherwise should do perfectly. Thanks for providing testing instructions too

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (JohnN193) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:28]: [q] should this be defined within the viam::utils namespace? Or is it okay to leave as a global

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (JohnN193) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:48]: [opt]we haven't been using block comments for other functions, not something that needs to change though

**Review by JohnN193 — APPROVED** [review]
Spoke offline about moving the helper functions into the `utils` namespace. otherwise just some small stuff, LGTM

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (nicksanford) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:60]: Thanks for the review y'all. Just to get another opinion I asked eric & he said that this is pretty much the only way to get each byte out of a float in C++ (which is not super safe no matter what one does) & is the same in terms of safety to the original.

As a result I'll keep this as it is a win for readability.

He did bring up 2, longer term (maybe for us to chat about later @tessavitabile) concerns, not with this implementation but more with the [PCD spec](https://pointclouds.org/documentation/tutorials/pcd_file_format.html) itself) which says:
`in binary form, where the data is a complete memory copy of the pcl::PointCloud.points array/vector. On Linux systems, we use mmap/munmap operations for the fastest possible read/write access to the data.`

What that means is that data corruption will likely silently occur if either of the following are true:
1.  the encoder & decoder processors have different endianness (i.e. not little endian)
2. the encoder & decoder processors have different floating point representations (i.e. not ieee 754)

Currently all our code assumes little endian (as that is what amd & intel support) & ieee 754.

This isn't relevant to this change but we may want to consider providing additional metadata in the PCD response regarding the floating point spec to use to interpret the bytes & what endianness to use as well, by checking on the PCD encoder side for these behaviors on boot.

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.h
Comment (nicksanford) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.h]: done

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.h
Comment (nicksanford) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.h]: done

**Pull request merged**
